isisd: Add support for SRv6 uSID (RFC 9352)

cscarpitta opened the pull request:
This PR implements the SRv6 uSID IS-IS extensions (RFC 9352). It extends the current IS-IS CLI to enable SRv6 on a specific IS-IS area and assign an already configured SRv6 locator to it.

```
router isis FOO
  net 49.0001.1111.1111.1111.00
  is-type level-2-only
  metric-style wide
  segment-routing srv6
   locator loc1
```

The SRv6 locator is configured in zebra using the below CLI.

```
segment-routing
 srv6
  locators
   locator loc1
    prefix fc00:0:1::/48 block-len 32 node-len 16 func-bits 16
    behavior usid
```

Once SRv6 is enabled and an SRv6 locator is configured, IS-IS advertises the SRv6 capabilities and the SRv6 locator to the IS-IS neighbors in the LSPs. Additionally, IS-IS automatically allocates an SRv6 uN SID (a.k.a Prefix SID) and a uA SID (a.k.a. Adjacency SID) for each IS-IS adjacency. The allocated SIDs are installed in the data plane through zebra and advertised to the IS-IS neighbors using the appropriate TLV messages according to the rules described in RFC 9352.

On the receiving side, when IS-IS receives an SRv6 Locator TLV, it computes a route to the locator using the SPF algorithm and installs the route locally in the RIB.

The following table summarizes the supported TLV and Sub-TLV messages:

Type | Reference | Status
-- | -- | --
SRv6 Capabilities Sub-TLV | [RFC 9352, Section 2](https://datatracker.ietf.org/doc/html/rfc9352#name-srv6-capabilities-sub-tlv) | Supported
Segment Routing Algorithm Sub-TLV | [RFC 8667, Section 3.2](https://datatracker.ietf.org/doc/html/rfc8667#name-sr-algorithm-sub-tlv) | Supported
Node MSD Sub-TLV | [RFC 8491, Section 2](https://datatracker.ietf.org/doc/html/rfc8667#name-sr-algorithm-sub-tlv) | Supported
SRv6 Locator TLV Format | [RFC 9352, Section 7.1](https://datatracker.ietf.org/doc/html/rfc9352#name-srv6-locator-tlv-format) | Supported
SRv6 End SID Sub-TLV | [RFC 9352, Section 7.2](https://datatracker.ietf.org/doc/html/rfc9352#name-srv6-end-sid-sub-tlv) | Supported
SRv6 End.X SID Sub-TLV | [RFC 9352, Section 8.1](https://datatracker.ietf.org/doc/html/rfc9352#name-srv6-endx-sid-sub-tlv) | Supported
SRv6 LAN End.X SID Sub-TLV | [RFC 9352, Section 8.2](https://datatracker.ietf.org/doc/html/rfc9352#name-srv6-lan-endx-sid-sub-tlv) | Supported
SRv6 SID Structure Sub-Sub-TLV | [RFC 9352, Section 9](https://datatracker.ietf.org/doc/html/rfc9352#name-srv6-sid-structure-sub-sub-) | Supported

We would like to thank Dmytro Shyty (6wind), Hiroki Shirokura (LINE Corporation) and Naoyuki Tachibana (LINE Corporation) for their input and contribution.

Signed-off-by: Carmine Scarpitta <carmine.scarpitta.94@gmail.com>

dmytroshytyi-6WIND reviewed (commented):
Hello,
I provide the first small set of comments - after first light overview. My comments are generally going to be related and limited to ISIS SRv6 functionality defined in RFC9352. (Possibly including SRv6-End, SRv6-End.X and LAN End.X). My comments are not related to uSID instructions.

cscarpitta commented on isisd/isis_srv6.c:
```diff
@@ -8,3 +8,22 @@
+ */
+void isis_srv6_area_init(struct isis_area *area)
+{
+	struct isis_srv6_db *srv6db = &area->srv6db;
```
Thanks for the comment. I will add a check to ensure that `area` is not NULL before accessing its members.

cscarpitta replied:
Fixed.

cscarpitta commented on isisd/isisd.h:
```diff
@@ -342,6 +342,7 @@ void config_end_lsp_generate(struct isis_area *area);
 /* YANG paths */
 #define ISIS_INSTANCE	"/frr-isisd:isis/instance"
 #define ISIS_SR		"/frr-isisd:isis/instance/segment-routing"
+#define ISIS_SRV6 "/frr-isisd:isis/instance/segment-routing-srv6"
```
I will fix the misalignment. Thanks!

cscarpitta replied:
Fixed.

dmytroshytyi-6WIND commented:
It would be great to have an IS-IS SRv6 topotest with more that 2 routers and example of `sharp` command to steer traffic manually through the nodes (bootstrapped by IS-IS Srv6).

dmytroshytyi-6WIND commented:
Hello, 
Considering the next RFC 9352, could you please explain the possible mismatch, if there is any, of `End` behavior defined by [RFC8986](https://www.rfc-editor.org/rfc/rfc8986.html#name-initial-registrations) and `End` behavior observed:

config:
```
router isis 1
 is-type level-1
 net 49.0000.0000.0000.0001.00
 lsp-timers gen-interval 2 refresh-interval 900 max-lifetime 1200
 topology ipv6-unicast
 segment-routing srv6
  locator loc1
exit
!
segment-routing
 srv6
  locators
   locator loc1
    prefix fc00:0:1::/48 block-len 32 node-len 16 func-bits 16
end
```

show
```
r1# show ipv6 route
Codes: K - kernel route, C - connected, S - static, R - RIPng,
       O - OSPFv3, I - IS-IS, B - BGP, N - NHRP, T - Table,
       A - Babel, D - SHARP, F - PBR, f - OpenFabric,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

[...]
I>* fc00:0:1::/128 [115/0] is directly connected, sr0, seg6local End USP, weight 1, 1d00h23m
```

If there are no such mismatch, it might be useful to introduce appropriate example-topotest.

cscarpitta commented on isisd/isis_cli.c:
```diff
@@ -1934,6 +1934,15 @@ void cli_show_isis_prefix_sid_algorithm(struct vty *vty,
+				bool show_defaults)
+{
+	if (!yang_dnode_get_bool(dnode, NULL))
+		vty_out(vty, " no");
```
Good catch! When we do 'no segment-routing srv6', we should unset/unadvertise the locator and uninstall/unadvertise all the SIDs, just like in BGP SRv6 and SR-MPLS IS-IS.

cscarpitta commented on isisd/isis_srv6.c:
```diff
@@ -0,0 +1,768 @@
+ *
+ * @result 0 on success, -1 otherwise
+ */
+int isis_srv6_locator_unset(struct isis_area *area)
```
Make sense. I will change to bool. Thanks!

cscarpitta replied:
Changed return value to bool.

cscarpitta commented on isisd/isis_srv6.h:
```diff
@@ -10,13 +10,31 @@
 #ifndef _FRR_ISIS_SRV6_H
 #define _FRR_ISIS_SRV6_H
 
+/* Maximum SRv6 SID Depths supported by the router */
```
Yes, I'm updating the PR to address all your comments. As discussed, I will add some CLI commands to customize these values.

cscarpitta commented on isisd/isis_srv6.c:
```diff
@@ -0,0 +1,768 @@
+			   ? SRV6_ENDPOINT_BEHAVIOR_END_X_NEXT_CSID
+			   : SRV6_ENDPOINT_BEHAVIOR_END_X;
+
+	sra = XCALLOC(MTYPE_ISIS_SRV6_INFO, sizeof(*sra));
```
It is deallocated in `srv6_endx_sid_del()`, which is called when

- an IS-IS adjacency goes down
- an IS-IS adjacency loses its IPv6 address
- SRv6 is turned off on an IS-IS area
- the locator is unset 

https://github.com/FRRouting/frr/blob/3b9417caff2b7bc872beedac1f3151cbf02a977b/isisd/isis_srv6.c#L505-L510

cscarpitta commented on isisd/isis_srv6.c:
```diff
@@ -0,0 +1,768 @@
+	struct isis_srv6_sid *sid = NULL;
+
+	if (!area || !chunk)
+		return false;
```
I will fix it, thanks!

cscarpitta replied:
Fixed.

dmytroshytyi-6WIND commented:
> It would be great to have an IS-IS SRv6 topotest with more that 2 routers and example of `sharp` command to steer traffic manually through the nodes (bootstrapped by IS-IS Srv6).
Possible example of the topology might be:
```r1 - r2 - r3 - r4 - r5```.

cscarpitta commented on doc/user/isisd.rst:
```diff
@@ -589,6 +589,40 @@ The following command show Flex-Algo information:
+Segment Routing over IPv6 (SRv6)
+================================
+
+This feature enables extensions in IS-IS to support Segment Routing with IPv6
```
Fixed.

cscarpitta commented on isisd/isis_lsp.c:
```diff
@@ -1210,6 +1210,57 @@ static void lsp_build(struct isis_lsp *lsp, struct isis_area *area)
+	}
+
+	/* Add SRv6 Locator TLV. */
+	if (!list_isempty(area->srv6db.srv6_locator_chunks)) {
```
Fixed.

cscarpitta commented on isisd/isis_lsp.c:
```diff
@@ -1210,6 +1210,57 @@ static void lsp_build(struct isis_lsp *lsp, struct isis_area *area)
+				srv6db->config.max_h_encaps_msd;
+			rcap->srv6_msd.max_end_d_msd =
+				srv6db->config.max_end_d_msd;
+		}
```
Fixed.

odd22 reviewed (commented):
Thanks for your great contribution. The code looks good to me in its overall. I have just some minor remarks:
- Can you check if it is possible to use loopback interface instead of dedicated sr0 interface like for SR-MPLS?
- Please replace strmatch by strncmp in various place.
- Can you fix the faulty test you introduce?

dmytroshytyi-6WIND reviewed (changes requested):
Please update the PR to pass the tests. Then I continue the review.

cscarpitta commented:
> Thanks for your great contribution. The code looks good to me in its overall. I have just some minor remarks:
> 
> * Can you check if it is possible to use loopback interface instead of dedicated sr0 interface like for SR-MPLS?
> * Please replace strmatch by strncmp in various place.
> * Can you fix the faulty test you introduce?

Hi @odd22, 

Many thanks for the review. 

The current Linux implementation of IPv6 Lightweight tunnel cannot use the loopback device. That’s the reason we decided to use dedicated interface (`sr0`). 

For example, when you install an SRv6 SID using the loopback device, the kernel ignores the SRv6 attributes:

```
# ip -6 route add fc00:0:1:100:: encap seg6local action End dev lo
```

```
# ip -6 route show
fc00:0:1:100:: dev lo metric 1024 pref medium
```


The same SRv6 SID is installed correctly (and the SRv6 attributes are present) when you use any non-loopback device:

```
# ip -6 route add fc00:0:1:100:: encap seg6local action End dev sr0
```

```
# ip -6 route show
fc00:0:1:100::  encap seg6local action End dev sr0 metric 1024 pref medium
```

---

As suggest, I also replaced `strmatch` by `strncmp` and fixed the failing Topotests.

cscarpitta commented:
ci:rerun unrelated failures

cscarpitta commented:
Rebased on master branch.

dmytroshytyi-6WIND commented on isisd/isis_tlvs.c:
```diff
@@ -1970,6 +2477,76 @@ static int pack_items_(uint16_t mtid, enum isis_tlv_context context,
+{
+	struct isis_subsubtlvs *result;
+
+	result = XCALLOC(MTYPE_ISIS_SUBSUBTLV, sizeof(*result));
```
There are 2 functions for END.X : `srv6_endx_sid_add` and `srv6_endx_sid_del`.

Here `isis_alloc_subsubtlvs` is called from `srv6_endx_sid_add`. I see here the memory allocation when END.X sid is added.

We also want to have XFREE of this allocated memory in `srv6_endx_sid_del` right?

riw777 replied:
we probably want to run this through some sort of memory leak check before committing, if possible ... lots of allocs and frees, we might won't catch all of them by inspection (?)

cscarpitta replied:
@riw777 Thanks for the suggestion. I used valgrind to discover possible memory leaks and I fixed all of them.

dmytroshytyi-6WIND commented on isisd/isis_tlvs.c:
```diff
@@ -6934,3 +8160,95 @@ void isis_tlvs_set_purge_originator(struct isis_tlvs *tlvs,
+		    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT4_USID &&
+		    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT46 &&
+		    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT46_USID) {
+			isis_subtlvs_add_srv6_end_sid(loc_tlv->subtlvs, sid);
```
Why do we call the `isis_subtlvs_add_srv6_end_sid()` when `sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END...`?

cscarpitta replied:
We should call `isis_subtlvs_add_srv6_end_sid()` only if `sid->behavior == SRV6_ENDPOINT_BEHAVIOR_END... ` Fixed.

dmytroshytyi-6WIND commented:
it seems you copied sr-ipv6 router capability format from r-mpls router capability format.
But it seems it is not quite the same.
I would suggest optimisations.

riw777 commented on isisd/isis_tlvs.c:
```diff
@@ -4366,6 +5153,75 @@ static int pack_tlv_router_cap(const struct isis_router_cap *router_cap,
 	}
 #endif /* ifndef FABRICD */
```
not specific to this code ... I'd really like to take fabricd out and add dist-opt-flood :-)

riw777 reviewed (commented):
some minor comments ... code looks pretty good ... we should push this through fuzzing and memleak checking (if not already done)

cscarpitta commented on isisd/isis_nb_config.c:
```diff
@@ -3365,6 +3366,118 @@ int isis_instance_flex_algo_priority_destroy(struct nb_cb_destroy_args *args)
+
+	loc_name = yang_dnode_get_string(args->dnode, NULL);
+
+	if (strmatch(loc_name, area->srv6db.config.srv6_locator_name)) {
```
Replaced with strncmp.

cscarpitta commented on isisd/isis_spf.c:
```diff
@@ -1128,6 +1128,51 @@ static int isis_spf_process_lsp(struct isis_spftree *spftree,
+			them. If we find the SRv6 Locator in some Prefix
+			Reachbility TLV then it means that we have already
+			processed it before and we can skip it. */
+			bool loc_is_in_ipv6_reach = false;
```
Moved to the top.

cscarpitta commented on isisd/isis_srv6.c:
```diff
@@ -0,0 +1,768 @@
+	struct isis_srv6_sid *sid;
+	struct srv6_adjacency *sra;
+
+	if (strmatch(area->srv6db.config.srv6_locator_name, "")) {
```
Replaced with strncmp.

cscarpitta commented on isisd/isis_srv6.h:
```diff
@@ -0,0 +1,178 @@
+
+/* Name of the interface used for installing SRv6 SIDs into the data plane */
+#define SRV6_IFNAME "sr0"
+
```
@odd22 @dmytroshytyi-6WIND @riw777 I added a new CLI to make the interface configurable: https://github.com/FRRouting/frr/pull/13724/commits/e09ec26ecbdaa91f79c28b19e30e1a089d222340

cscarpitta commented:
Hi @riw777, many thanks for the review.

I addressed all comments, added a new CLI command to make the SRv6 interface configurable, and fixed all memory leaks identified by valgrind.

Could you please take another look?

riw777 reviewed (approved):
looks good

cscarpitta commented:
> It would be great to have an IS-IS SRv6 topotest with more that 2 routers and example of `sharp` command to steer traffic manually through the nodes (bootstrapped by IS-IS Srv6).

@dmytroshytyi-6WIND I changed the topotest as suggested, thanks.

ci:rerun CI failures are not related to the changes. Rerunning!

cscarpitta commented:
Hi @donaldsharp, @odd22, @riw777,  
 
I have addressed all the review comments received for this PR.
The PR has passed the CI test.
 
The PR should be ready for merge. Could you please have a look?
If you don’t have further comments, could we merge this before the code freeze of Sept 19?

dmytroshytyi-6WIND commented on isisd/isis_tlvs.c:
```diff
@@ -4227,6 +4227,21 @@ static size_t isis_router_cap_tlv_size(const struct isis_router_cap *router_cap)
 #endif /* ifndef FABRICD */
 
+	if (router_cap->srv6_cap.is_srv6_capable) {
+		sz += 2 + ISIS_SUBTLV_SRV6_CAPABILITIES_SIZE;
```
Do we want to use a define here for a magic value?

cscarpitta replied:
Fixed, thanks for the suggestion.

dmytroshytyi-6WIND commented on isisd/isis_tlvs.c:
```diff
@@ -4227,6 +4227,21 @@ static size_t isis_router_cap_tlv_size(const struct isis_router_cap *router_cap)
+
+		nb_algo = isis_tlvs_sr_algo_count(router_cap);
+		if (nb_algo != 0)
+			sz += 2 + nb_algo;
```
Do we want to use a define here for a magic value?

cscarpitta replied:
Fixed.

dmytroshytyi-6WIND commented on isisd/isis_tlvs.c:
```diff
@@ -4227,6 +4227,21 @@ static size_t isis_router_cap_tlv_size(const struct isis_router_cap *router_cap)
+			 router_cap->srv6_msd.max_h_encaps_msd +
+			 router_cap->srv6_msd.max_end_d_msd;
+		if (nb_msd != 0)
+			sz += 2 + 2 * nb_msd;
```
Do we want to use a define here for a magic value?

cscarpitta replied:
Fixed.

dmytroshytyi-6WIND commented on isisd/isis_tlvs.c:
```diff
@@ -2120,6 +2120,62 @@ static int pack_item_srv6_end_sid(struct isis_item *i, struct stream *s,
+		sbuf_push(
+			log, indent,
+			"Expected %hhu bytes of Sub-Sub-TLVs, but only %u bytes available.\n",
+			subsubtlv_len, len - 20);
```
Magic value... define?

cscarpitta replied:
Fixed.

dmytroshytyi-6WIND commented on isisd/isis_tlvs.c:
```diff
@@ -1340,6 +1342,87 @@ static int pack_item_ext_subtlvs(struct isis_ext_subtlvs *exts,
+			} else {
+				/* No Sub-Sub-TLVs */
+				if (STREAM_WRITEABLE(s) < 1) {
+					*min_len = 21;
```
Magic number "21", do we want to define it?

cscarpitta replied:
Fixed.

dmytroshytyi-6WIND commented on isisd/isis_tlvs.c:
```diff
@@ -1340,6 +1342,87 @@ static int pack_item_ext_subtlvs(struct isis_ext_subtlvs *exts,
+			} else {
+				/* No Sub-Sub-TLVs */
+				if (STREAM_WRITEABLE(s) < 1) {
+					*min_len = 27;
```
Magic munber "27"...

cscarpitta replied:
Fixed.

dmytroshytyi-6WIND reviewed (commented):
Hello @cscarpitta , thank you for addressing the comments.
I have a few more suggestions on values that used in the implementation and it probably would be nice to have them explained in the code. For example with `#define`.

dmytroshytyi-6WIND commented on doc/user/isisd.rst:
```diff
@@ -594,6 +594,40 @@ The following command show Flex-Algo information:
+   SID and adjacency SIDs, creates local SID entries in the data plane, and
+   advertises them in the IGP domain.
+
+.. clicmd:: interface NAME
```
I would expect `interface Name` under zebra to make this interface available not only for ISIS, but for other protocols?

dmytroshytyi-6WIND reviewed (approved):
Thank you for this great contribution!
LGTM.
